README.md
# P2P Manager
Manage a network of peers in a peer-to-peer network.

## Messages

### `peerConnect`, `peerEnd`, `peerError`
Emitted when an active Peer emits a `connect`, `end` or `error` message respectively, and the data payload is the same:

```
{
  peer: Peer
}
```

### `peerMessage`
Emitted when an active Peer sends a `message` event.

```
{
  peer: Peer,
  command: String,
  data: Raw payload as binary data
}
```

### `commandMessage`
An alternate version of the `peerMessage` event; in these events, the command of the message is used as the event name (i.e. command `'foo'` would cause a `fooMessage` event).

```
{
  peer: Peer,
  data: Raw payload as binary data
}
```

### `error`
Error message from the PeerManager. The `severity` attribute is one of 'info', 'notice', 'warning', or 'error' (in increasing severity). 

```
{
  severity: String,
  message: String
}
```

## `PeerManager.send()`
The PeerManager allows sending messages to a collection of Peers at once, based on certain criteria. The most common criteria is the "state" of the Peer. Each Peer object has a `state` property which is set to `'new'` when first created, `'connecting'` when first opened, and `'connected'` when first `connect` event is heard (before `handleConnect` is called, so that can be overwritten, if desired). Any other states are up to your PeerManager instance to implement.

* `number`: How many peers to send to? Once filtered down, should all or a sub-set be messaged? Pass a Number to send to that many Peers (picked at random). Passing a zero or negative number for this argument sends to all matched Peers.
* `prop`: (String) Which property of the Peers should be examined?
* `value`: Value of the property the Peers need to match to be included in the set. Matching is done in a case-insensitive way for strings. Passing an array for this argument requires Peers match any one of the array's values.
* `cmd`: (String) Message name to be sent to the Peers
* `payload`: (Buffer) Binary data to be sent to the Peers
* `answerCmd`: (String) Message name of answering message. If null/false, defaults to all `message` events.
* `callback`: (function) If provided, is bound to the "`answerCmd`" event of the Peers contacted, to await their reply

```js
PeerManager.send('all', 'state', 'connected', 'hi', new Buffer([1,2,3,4,5])); // Send a message to all connected clients
PeerManager.send(2, 'state', 'lonely', 'matchmaker', new Buffer([1,2,3,4,5])); // Send a message to a random two Peers who have state=='lonely'
PeerManager.send(5, 'myProp', [1,5,42,false], 'hi', new Buffer([1,2,3,4,5])); // Send a message to a random five Peers who have myProp equal to either 1, 5 ,42, or false
```

The function returns an dictionary of Peers the message was sent to, stored by their UUID.

If you're expecting a specific answer to your message, there's a few ways you can listen in:

### Listen to only answer messages, from just the filtered peers
Probably the most common, and the most specific listening type:

```js
var m = new PeerManager();
var waitForAnswer = function(d) {
  console.log(d.peer.getUUID()+': has answered', d.data.toString('hex'));
  d.peer.removeListener('message', this); // This listener doesn't care about further messages
  delete peers[d.peer.getUUID()]; // Remove this peer from the list of peers who haven't answered yet
};

var peers = m.send(2, 'state', 'waiting', 'knock', new Buffer([1,2,3]), 'answer', waitForAnswer);

setTimeout(function() {
  // Ten seconds have passed; give up on those who haven't answered
  for (var uuid in peers) {
    if (peers.hasOwnProperty(uuid)) {
      console.log(uuid+' never answered...');
      peers[uuid].removeListener('message', waitForAnswer);
    }
  }
}, 10000);
```

### Listen to all messages, from just the filtered Peers 
A good method if you expect the answer you seek to be in the next few messages from those Peers, and the answer might be one of several messages ('answer' or 'error'):

```js
var m = new PeerManager();
var waitForAnswer = function(d) {
  if (d.command !== 'answer') return; // Wait for next message...
  
  console.log(d.peer.getUUID()+': has answered', d.data.toString('hex'));
  d.peer.removeListener('message', this); // This listener doesn't care about further messages
  delete peers[d.peer.getUUID()]; // Remove this peer from the list of peers who haven't answered yet
};

var peers = m.send(2, 'state', 'waiting', 'knock', new Buffer([1,2,3]), false, waitForAnswer);

setTimeout(function() {
  // Ten seconds have passed; give up on those who haven't answered
  for (var uuid in peers) {
    if (peers.hasOwnProperty(uuid)) {
      console.log(uuid+' never answered...');
      peers[uuid].removeListener('message', waitForAnswer);
    }
  }
}, 10000);
```

### Listen to only answer messages, from all Peers
If there are lots of other messages being sent around, but very few of the particular answer messages you're looking for, this works well, or if there's a possibility another peer will answer on behalf of the peer you requested from:

```js
var m = new PeerManager();
var waitForAnswer = function(d) {
  // Look through our list of peers and see if d.peer is one of them
  if (peers[d.peer.getUUID()] !== d.peer) return;
  
  console.log(d.peer.getUUID()+': has answered', d.data.toString('hex'));
  delete peers[d.peer.getUUID()]; // Remove this peer from the list of peers who haven't answered yet
};
var peers = m.send(2, 'state', 'waiting', 'knock', new Buffer([1,2,3]));
m.on('answer', waitForAnswer); // Listen in on all 'answer' messages from all peers

setTimeout(function() {
  // Ten seconds have passed; give up on those who haven't answered
  for (var uuid in peers) {
    if (peers.hasOwnProperty(uuid)) {
      console.log(uuid+' never answered...');
    }
  }
  m.removeListener('answer', waitForAnswer);
}, 10000);
```

## Peers
The PeerManager keeps a list of known Peers to connect to (the "pool"), as well as a list of those currently connected ("active").

As your implementation discovers new Peers, use the `PeerManager.addPool(hosts)` method to tell the PeerManager about them. If the number of active Peers is currently below the minimum (`options.minPeers`), a connection will be attempted immediately. Otherwise, they will just be added to the pool. `PeerManager.addActive(hosts)` adds a new peer and attempts to connect to it immediately, regardless if the number of active Peers is above the threshhold.

## Peer lists
The `seedPeers` argument of `launch()`, and the `hosts` argument of `addPool()` and `addActive()` are lists of hosts, expressed one of the following ways:

* If it's a String, it's assumed to be the hostname/ip of one host to use; make sure to include a port number as part of the hostname (colon-delimited).
* If it's an Array, it's looped through and each element inspected.
* * If it's a String, it's assumed to be a hostname/ip of a host to use.
* * If it's an Array, index zero is used as host and index one is used as port.
* * If it's an Object, the `host` property is used as host, and the `port` property is used as port.


CHANGELOG.md
Next Release
------------------
* Use `Object.setProperty()` accessor functions for emitting state changes

0.1.0 / 2014-01-10
------------------
* Initial release

test/manager.test.js
var PeerManager = require('../lib/PeerManager').PeerManager;
var Peer = require('p2p-node').Peer;
var assert = require("assert");
var net = require('net');

describe('P2P PeerManager', function() {
  it('should properly connect to indicated host', function(done) {
    var localManager = new PeerManager({listen:false});
    localManager.on('error', function(err) {
      if (err.severity !== 'info' && err.severity !== 'notice') {
        console.log(err);
        assert.ok(false);
      }
    });
    var server = net.createServer(function(socket) {
      server.close();
      localManager.shutdown();
      done();
    });
    server.listen(function() {
      localManager.launch([{host:server.address().address, port:server.address().port}]);
    });
  });
  describe('Messaging', function() {
    var magic = 0x01020304;
    var server = false;
    var localManager = false;
    var serverPeer = false;
    
    beforeEach(function(done) {
      serverPeer = false;
      server = net.createServer(function(socket) {
        serverPeer = new Peer(socket.remoteAddress, socket.remotePort, magic);
        serverPeer.connect(socket);
        done();
      });
      localManager = new PeerManager({listen: false, magic: magic});
      localManager.on('error', function(err) {
        //console.log(err.severity+': '+err.message);
      });
      server.listen(function() {
        localManager.launch([{host:server.address().address, port:server.address().port}]);
      });
    });
    
    afterEach(function() {
      if (serverPeer !== false) serverPeer.destroy();
      server.close();
      if (localManager !== false) localManager.shutdown();
    });
  
    it('should bubble up message events from peers', function(done) {
      var timer = false;
      localManager.on('error', function(err) {
        if (err.severity !== 'info' && err.severity !== 'notice') {
          assert.ok(false);
        }
      });
      localManager.once('message', function(d) {
        assert.equal(d.command, 'hello');
        assert. equal(d.data.toString('utf8'), 'world');
        clearInterval(timer);
        done();
      });
      timer = setInterval(function() {
        serverPeer.send('hello', new Buffer('world', 'utf8'));
      }, 100);
    });
    it('should bubble up command message events from peers', function(done) {
      var timer = false;
      localManager.on('error', function(err) {
        if (err.severity !== 'info' && err.severity !== 'notice') {
          assert.ok(false);
        }
      });
      localManager.once('helloMessage', function(d) {
        assert.equal(d.data.toString('utf8'), 'world');
        clearInterval(timer);
        done();
      });
      timer = setInterval(function() {
        serverPeer.send('hello', new Buffer('world', 'utf8'));
      }, 100);
    });
    it('should send messages to peers', function(done) {
      var timer = false;
      localManager.on('error', function(err) {
        if (err.severity !== 'info' && err.severity !== 'notice') {
          assert.ok(false);
        }
      });
      serverPeer.once('helloMessage', function(d) {
        assert.equal(d.data.toString('utf8'), 'world');
        clearInterval(timer);
        done();
      });
      timer = setInterval(function() {
        localManager.send('all', 'state', 'connected', 'hello', new Buffer('world', 'utf8'));
      }, 100);
    });
  });
  describe('Server listening', function() {
    var magic = 0x01020304;
    var port = 8080;
    var managerServer = new PeerManager({listen:true, magic: magic, port: port});
    var localPeer = false;
    
    managerServer.on('error', function(err) {
      //console.log(err.severity+': '+err.message);
    });
    managerServer.launch([]);

    afterEach(function() {
      if (localPeer !== false) localPeer.destroy();
    });
    
    it('should listen for new peers', function(done) {
      managerServer.once('listenConnect', function(p) {
        done();
      });
      localPeer = new Peer('127.0.0.1', port, magic);
      localPeer.connect();
    });
    it('should receive messages from remote peers', function(done) {
      managerServer.once('helloMessage', function(d) {
        assert.equal(d.data.toString('utf8'), 'world');
        clearInterval(timer);
        done();
      });
      localPeer = new Peer('127.0.0.1', port, magic);
      localPeer.connect();
      timer = setInterval(function() {
        localPeer.send('hello', new Buffer('world', 'utf8'));
      }, 100);
    });
  });
});


lib/Manager.js
var events = require('events');
var util = require('util');
var Peer = require('crypto-p2p-node').Peer;

var Manager = exports.Manager = function Manager(options) {
  events.EventEmitter.call(this);

  options = (typeof options === 'undefined')? {} : options;
  var defaultOpts = {
    'useCache': true,
    'listen': true,
    'port': 8333,
    'magic': 0xD9B4BEF9,
    'minPeers': 3,
    'maxPeers': 20
  };
  for (var name in defaultOpts) {
    if (defaultOpts.hasOwnProperty(name) && !options.hasOwnProperty(name)) {
      options[name] = defaultOpts[name];
    }
  }
  options.port = parseInt(options.port);
  options.magic = parseInt(options.magic);
  options.minPeers = parseInt(options.minPeers);
  options.maxPeers = parseInt(options.maxPeers);
  
  this.options = options;
  
  this.activePeers = {};
  this.activePeerCount = 0;
  
  this.poolPeers = [];
  
  this.state = 'new';
}
util.inherits(Manager, events.EventEmitter);

Manager.prototype.launch = function launch(seedPeers) {
  this.state = 'launching';
  
  if (seedPeers !== false) this.addActive(seedPeers); // Open connections to seed peers
  var self = this;
  setImmediate(function() { self.fillActive(); }); // Attempt to fill remaining peers from pool cache and start timer
  
  this.state = 'running';
};

Manager.prototype._parseHostList = function _parseHostList(elem) {
  var host, port;
  if (typeof elem == 'string') {
    host = elem;
    port = this.options.port;
  } else if (Array.isArray(elem)) {
    host = elem[0];
    port = (elem.length > 1)? elem[1] : this.options.port;
  } else if (typeof elem == 'object' && typeof elem.host !== 'undefined') {
    host = elem.host;
    port = (typeof elem.port !== 'undefined')? elem.port : this.options.port;
  } else {
    return false;
  }
  return [host, port];
};

// Add a new peer to the pool.
// If the number of active peers is below the threshhold, connect to them immediately.
Manager.prototype.addPool = function addPool(hosts) {
  if (typeof hosts == 'string') {
    hosts = [hosts];
  }
  for (var i = 0; i < hosts.length; i++) {
    var rs = this._parseHostList(hosts[i]);
    if (rs === false) continue;
    
    if (this.activePeerCount < this.options.minPeers) {
      this.addActive(rs[0], rs[1]);
    } else {
      this.poolPeers.push({host: rs[0], port: rs[1]});
    }
  }
  
  // De-duplicate poolPeers
  var unique = {};
  var distinct = [];
  this.poolPeers.forEach(function (peer) {
    var id = peer.host+'~'+peer.port;
    if (!unique[id]) {
      distinct.push(peer);
      unique[id] = true;
    }
  });
  this.poolPeers = distinct;
  
  return true;
}

Manager.prototype.addActive = function addActive(hosts) {
  if (typeof hosts == 'string') {
    hosts = [hosts];
  }
  for (var i = 0; i < hosts.length; i++) {
    var rs = this._parseHostList(hosts[i]);
    if (rs === false) continue;

    var p = this._connect(rs[0], rs[1]);
    this.activePeers[p.getUUID()] = p;
    this.activePeerCount++;
  }
}

// Internal function; don't call directly. Use addPool or addActive instead.
Manager.prototype._connect = function _connect(host, port) {
  port = (typeof port === 'undefined')? this.options.port : port;
  var p = new Peer(host, port, this.options.magic);
  p.state = 'new';
  
  self = this;
  p.once('connect', function(d) {
    console.log(d.peer.host.host+': connect');
    console.log(d.peer.host.host+' resolved to '+d.peer.socket.remoteAddress+':'+d.peer.socket.remotePort);
    d.peer.state = 'connected';
    
    if (typeof self.handleConnect === 'function') {
      var rs = self.handleConnect(d.peer);
      if (rs === false) self._disconnect(d.peer);
    }
  });
  p.once('end', function(d) {
    console.log(d.peer.host.host+': end');

    if (typeof self.handleEnd === 'function') {
      var rs = self.handleEnd(d.peer);
    }
    if (d.peer.state !== 'disconnecting') self._disconnect(d.peer); // Other end hung up on us; no need to hang around
  });
  p.on('error', function(d) {
    console.log(d.peer.host.host+': error', d.error);

    if (typeof self.handleError === 'function') {
      self.handleError(d.peer);
    }
    if (d.peer.state !== 'disconnecting') self._disconnect(d.peer); // Close the connection that errored
  });
  p.on('message', function(d) {
    console.log(d.peer.host.host+': message', d.command, d.data.toString('hex'));
    for (var prop in self) {
      if (self.hasOwnProperty(prop) && typeof self[prop] == 'function' && prop == d.command+'Message') {
        var rs = self[prop](d.data, d.peer);
        if (rs === false) {
          // Stop execution
          return;
        } else {
          // Also trigger an event
          break;
        }
      }
    }
    
    // No handler found; send it as a message.
    this.emit(d.command, {
      peer: d.peer,
      data: d.data
    });
  });

  setImmediate(function() {
    p.connect();
    p.state = 'connecting';
  });
  return p; // delay p.connect() using setImmediate, so that whatever is receiving this return value can prepare for the connection before it happens
};

Manager.prototype.fillActive = function fillActive() {
  if (this.activePeerCount >= this.options.minPeers) {
    // No action needed now; delay for a minute and check again
    clearTimeout(this.fillTimeout);
    this.fillTimeout = setTimeout(this.fillActive, 60*1000);
    this.fillTimeout.unref(); // If this timer is the only thing going, don't keep program open just for it
    return;
  }
  console.log('Too few active peers ('+this.activePeerCount+' < '+this.options.minPeers+'); pulling more from pool');

  while (this.activePeerCount < this.options.minPeers) {
    if (this.poolPeers.length == 0) {
      console.log('No more pooled peers...');
      return;
    }
    
    var peer = this.poolPeers.shift();
    this.addActive(peer.host, peer.port); // Synchronously queues up a Peer to be connected, and increments activePeerCount, so this while loop works
  }
};

Manager.prototype.delActive = function delActive(p) {
  delete this.activePeers[p.getUUID()];
  this.activePeerCount--;
  
  if (this.state = 'shutdown') return; // Don't attempt to add more peers if in shutdown mode
  var self = this;
  setImmediate(function() { self.fillActive(); });
};

// Internal function; don't call directly. Use delActive instead.
Manager.prototype._disconnect = function _disconnect(p) {
  var self = this;
  p.state = 'disconnecting';
  p.once('close', function(d) {
    console.log(d.peer.host.host+' is now closed');
    self.delActive(d.peer);
  });
  p.disconnect();
};

Manager.prototype.shutdown = function shutdown() {
  this.state = 'shutdown';
  for (var uuid in this.activePeers) {
    if (this.activePeers.hasOwnProperty(uuid) && this.activePeers[uuid] instanceof Peer) {
      console.log('Disconnecting '+uuid);
      this._disconnect(this.activePeers[uuid]);
    }
  }
};


lib/PeerManager.js
var events = require('events');
var net = require('net');
var util = require('util');
var Peer = require('p2p-node').Peer;

var PeerManager = exports.PeerManager = function PeerManager(options) {
  events.EventEmitter.call(this);

  options = (typeof options === 'undefined')? {} : options;
  var defaultOpts = {
    'useCache': true,
    'listen': true,
    'port': 8333,
    'magic': 0xD9B4BEF9,
    'minPeers': 3,
    'maxPeers': 20,
    'idleTimeout': 30*60*1000 // time out peers we haven't heard anything from in 30 minutes
  };
  for (var name in defaultOpts) {
    if (defaultOpts.hasOwnProperty(name) && !options.hasOwnProperty(name)) {
      options[name] = defaultOpts[name];
    }
  }
  options.port = parseInt(options.port);
  options.magic = parseInt(options.magic);
  options.minPeers = parseInt(options.minPeers);
  options.maxPeers = parseInt(options.maxPeers);
  
  this.options = options;
  
  this.activePeers = {};
  this.activePeerCount = 0;
  this.server = false;
  
  this.poolPeers = [];
  this.badPeers = {};
  
  var myState = 'new';
  Object.defineProperty(this, 'state', {
    enumerable: true,
    get: function() {
      return myState;
    },
    set: function(newValue) {
      var oldState = myState;
      this.emit('stateChange', {new: newValue, old: oldState});
      myState = newValue;
    }
  });
  
  return this;
}
util.inherits(PeerManager, events.EventEmitter);

PeerManager.prototype.launch = function launch(seedPeers) {
  this.state = 'launching';
  
  if (seedPeers !== false) this.addPool(seedPeers); // Open connections to seed peers
  if (this.options.listen) this.openListener(); // Listen for incoming connections
  
  var self = this;
  setImmediate(function() { self.checkPeers(); }); // Attempt to fill remaining peers from pool cache and start timer
  setInterval(function() { self.status(); }, 10*1000).unref(); // Send status message once a minute
};

PeerManager.prototype.openListener = function openListener() {
  var self = this;
  self._error('Opening listening port...', 'info');
  var server = this.server = net.createServer(function(socket) {
    var p = new Peer(socket.remoteAddress, socket.remotePort, self.options.magic);
    
    if (typeof self.activePeers[p.getUUID()] != 'undefined') {
      self._error('Already connected to peer '+p.getUUID()+' yet they contacted us?', 'warning');
      return;
    }
    
    p.on('end', function(d) {
      self.emit('peerEnd', d); // bubble up!
      if (d.peer.state !== 'disconnecting') self._disconnect(d.peer); // Other end hung up on us; no need to hang around
    });
    p.on('error', function(d) {
      self.emit('peerError', d); // bubble up!
      self._warn(d.peer, 10);
      if (d.peer.state !== 'disconnecting') self._disconnect(d.peer); // Close the connection that errored
    });
    p.on('message', function(d) {
      self.emit('message', d); // bubble up!
      self.emit(d.command+'Message', {
        peer: d.peer,
        data: d.data
      });
    });

    p.connect(socket); // Bind to existing socket
    self.activePeers[p.getUUID()] = p;
    self.activePeerCount++;

    if (self.state = 'launching') {
      self.state = 'running'; // First peer to connect updates state
    }
    self.emit('listenConnect', { peer:p });
  });

  server.on('error', function (e) {
    if (e.code == 'EADDRINUSE') {
      self._error('Can\'t open listening on port '+self.options.port+'; port already in use', 'warning');
    }
  });
  server.listen(this.options.port, function() {
    self._error('Now accepting connections on port '+self.options.port, 'info');
  });
};

PeerManager.prototype._parseHostList = function _parseHostList(elem) {
  var host, port;
  if (typeof elem == 'string') {
    host = elem;
    port = this.options.port;
  } else if (Array.isArray(elem)) {
    host = elem[0];
    port = (elem.length > 1)? elem[1] : this.options.port;
  } else if (typeof elem == 'object' && typeof elem.host !== 'undefined') {
    host = elem.host;
    port = (typeof elem.port !== 'undefined')? elem.port : this.options.port;
  } else {
    return false;
  }
  // Check for bad peer
  var uuid = host+'~'+port;
  if (typeof this.badPeers[uuid] != 'undefined') {
    // Warning exists; check and see if it's expired
    if ((new Date().getTime() - self.badPeers[uuid].date)/1000/60 > this.badPeers[uuid].warning) {
      delete self.badPeers[uuid]; // Warning expired
    } else {
      return false; // Warning still active
    }
  }
  return [host, port];
};

// Add a new peer to the pool.
// If the number of active peers is below the threshhold, connect to them immediately.
PeerManager.prototype.addPool = function addPool(hosts) {
  if (typeof hosts == 'string') {
    hosts = [hosts];
  }
  var target = (this.state == 'launching')? (this.options.maxPeers + this.options.minPeers)/2 : this.options.minPeers;
  for (var i = 0; i < hosts.length; i++) {
    var rs = this._parseHostList(hosts[i]);
    if (rs === false) continue;
    
    
    if (this.activePeerCount < target) {
      this.addActive([rs]);
    } else {
      if (typeof this.activePeers[rs[0]+'~'+rs[1]] !== 'undefined') {
        // Already connected to this peer
        continue;
      }
      this.poolPeers.push({host: rs[0], port: rs[1]});
    }
  }
  
  // De-duplicate poolPeers
  var unique = {};
  var distinct = [];
  this.poolPeers.forEach(function (peer) {
    var id = peer.host+'~'+peer.port;
    if (!unique[id]) {
      distinct.push(peer);
      unique[id] = true;
    }
  });
  this.poolPeers = distinct;
  
  return true;
}

PeerManager.prototype.addActive = function addActive(hosts) {
  if (typeof hosts == 'string') {
    hosts = [hosts];
  }
  for (var i = 0; i < hosts.length; i++) {
    var rs = this._parseHostList(hosts[i]);
    if (rs === false) continue;
    
    if (typeof this.activePeers[rs[0]+'~'+rs[1]] !== 'undefined') {
      // Already connected to this peer
      continue;
    }

    var p = this._connect(rs[0], rs[1]);
    this.activePeers[p.getUUID()] = p;
    this.activePeerCount++;
  }
}

// Internal function; don't call directly. Use addPool or addActive instead.
PeerManager.prototype._connect = function _connect(host, port) {
  port = (typeof port === 'undefined')? this.options.port : port;
  var p = new Peer(host, port, this.options.magic);
  
  self = this;
  p.on('connect', function(d) {
    //console.log(d.peer.host.host+' resolved to '+d.peer.socket.remoteAddress+':'+d.peer.socket.remotePort);
    if (self.state = 'launching') {
      self.state = 'running'; // First peer to connect updates state
    }
    self.emit('peerConnect', d); // bubble up!
  });
  p.on('end', function(d) {
    self.emit('peerEnd', d); // bubble up!
    if (d.peer.state !== 'disconnecting') self._disconnect(d.peer); // Other end hung up on us; no need to hang around
  });
  p.on('error', function(d) {
    self.emit('peerError', d); // bubble up!
    self._warn(d.peer, 10);
    if (d.peer.state !== 'disconnecting') self._disconnect(d.peer); // Close the connection that errored
  });
  p.on('message', function(d) {
    self.emit('message', d); // bubble up!
    self.emit(d.command+'Message', {
      peer: d.peer,
      data: d.data
    });
  });

  setImmediate(function() {
    self._error('Attempting to connect to '+p.getUUID(), 'notice');
    p.connect();
    setTimeout(function() { // Give them a few seconds to respond, otherwise close the connection automatically
      if (p.state == 'connecting') {
        p.state = 'closed';
        self._warn(p, 10);
        self.delActive(p, 'didn\'t respond to connection attempt; force-closing');
      }
    }, 5*1000).unref();
  });
  return p; // delay p.connect() using setImmediate, so that whatever is receiving this return value can prepare for the connection before it happens
};

// Periodic function that checks the status of the peer network being managed
PeerManager.prototype.checkPeers = function checkPeers() {
  if (this.state == 'shutdown') return;
  
  // First reset our own timer
  var self = this;
  clearTimeout(this.fillTimeout); // If we were called early, reset the existing one
  this.fillTimeout = setTimeout(function() { self.checkPeers(); }, 60*1000);
  this.fillTimeout.unref(); // If this timer is the only thing going, don't keep program open just for it
  
  // Timeout peers that have been quiet for too long
  for (var uuid in this.activePeers) {
    if (this.activePeers.hasOwnProperty(uuid) && this.activePeers[uuid].lastSeen !== false && this.activePeers[uuid].lastSeen.getTime() < new Date().getTime() - this.options.idleTimeout) {
      this._error(uuid+' has been quiet too long; disconnecting', 'info');
      this._disconnect(this.activePeers[uuid]);
    }
  }
  
  // Ensure minimum number of active peers are set
  if (this.activePeerCount < this.options.minPeers) {
    this._error('Too few active peers ('+this.activePeerCount+' < '+this.options.minPeers+'); pulling more from pool', 'info');
    while (this.activePeerCount < this.options.minPeers) {
      if (this.poolPeers.length == 0) {
        this._error('No more pooled peers...', 'info');
        return;
      }
    
      var peer = this.poolPeers.shift();
      this.addActive([peer]); // Synchronously queues up a Peer to be connected, and increments activePeerCount, so this while loop works
    }
  }
  
  // Warn if more than maximum is set
  if (this.activePeerCount > this.options.maxPeers) {
    this._error('Number of active peers above the maximum ('+this.activePeerCount+' > '+this.options.maxPeers+'); find a way to determine which should be disconnected, and call delActive() on them.');
  }
};

PeerManager.prototype.status = function status() {
  this.emit('status', {
    'numActive': this.activePeerCount,
    'poolSize': this.poolPeers.length,
    'badPeers': this.badPeers
  });
};

PeerManager.prototype.send = function send(number, property, values, cmd, payload, answer, callback) {
  if (number == false || number < 0) number = 'all';
  if (!Array.isArray(values)) values = [values];
  if (typeof payload == 'undefined') payload = new Buffer(0);

  // Build a sub-set of items
  var uuids = [];
  for (var uuid in this.activePeers) {
    if (this.activePeers.hasOwnProperty(uuid)) {
      for (var i = 0; i < values.length; i++) {
        if (this.activePeers[uuid][property] == values[i]) {
          uuids.push(uuid);
          break;
        }
      }
    }
  }
  if (uuids.length == 0) return false; // None matched that filter
  
  // Shuffle
  var i = uuids.length, randomIndex, temp;
  while (0 !== i) {
    // Pick a remaining element...
    randomIndex = Math.floor(Math.random() * i);
    i--;

    // And swap it with the current element.
    temp = uuids[i];
    uuids[i] = uuids[randomIndex];
    uuids[randomIndex] = temp;
  }
  
  var toSend = {};
  if (number == 'all') number = uuids.length;
  var picked = 0;
  while (picked < number) {
    var uuid = uuids.pop();
    var p = this.activePeers[uuid];
    if (typeof p !== 'undefined') { // Ensure it hasn't closed in the gap
      toSend[uuid] = p;
      picked++;
    }
    if (uuids.length == 0) break; // No more to parse
  }
  
  // now send to toSend:
  for (var uuid in toSend) {
    if (toSend.hasOwnProperty(uuid)) {
      toSend[uuid].send(cmd, payload);
      if (typeof callback == 'function') {
        toSend[uuid].on(answer, callback);
      }
    }
  }
  return toSend;  
};

PeerManager.prototype.delActive = function delActive(p, reason) {
  var self = this;
  if (typeof this.activePeers[p.getUUID()] == 'undefined') return; // We weren't connected to them in the first place
  if (p.state != 'disconnecting' && p.state != 'disconnected' && p.state != 'closed') {
    if (typeof reason == 'string') {
      this._error('Disconnecting from '+p.getUUID()+' ('+p.state+'); '+reason, 'info');
    } else {
      this._error('Disconnecting from '+p.getUUID()+' ('+p.state+')', 'info');
    }
    return this._disconnect(p); // Hang up first, then delete
  }
  
  p.destroy();
  delete this.activePeers[p.getUUID()];
  this.activePeerCount--;
  if (typeof reason == 'string') {
    this._error(p.getUUID()+' closed; '+reason+'. '+this.activePeerCount+' active peers now', 'info');
  } else {
    this._error(p.getUUID()+' closed; '+this.activePeerCount+' active peers now', 'info');
  }
  
  if (this.state == 'shutdown') return; // Don't attempt to add more peers if in shutdown mode
  setImmediate(function() { self.checkPeers(); });
};

PeerManager.prototype._warn = function _warn(p, mins) {
  if (typeof this.badPeers[p.getUUID()] != 'undefined') {
    // Already has a warning; see if it's expired
    var warningRemaining = this.badPeers[p.getUUID()].warning - ((new Date().getTime() - this.badPeers[p.getUUID()].date)/1000/60);
    if (warningRemaining < 0) warningRemaining = 0;
    // Add new warning to the old
    this.badPeers[p.getUUID()].warning = warningRemaining + mins;
    this.badPeers[p.getUUID()].date = new Date().getTime();
  } else {
    this.badPeers[p.getUUID()] = {
      host: p.host.host,
      port: p.host.port,
      date: new Date().getTime(),
      warning: mins
    };
  }
};

// Internal function; don't call directly. Use delActive instead.
PeerManager.prototype._disconnect = function _disconnect(p) {
  var self = this;
  p.state = 'disconnecting';
  p.once('close', function(d) {
    self.delActive(d.peer, 'remote connection closed');
  });
  p.disconnect();
  // Give them a few seconds to close out, otherwise we'll just delete
  setTimeout(function() {
    if (self.activePeers[p.getUUID()] == p) {
      // Hasn't been deleted yet
      self._error(p.getUUID()+' didn\'t close on their own; force-closing', 'notice');
      p.destroy(); // Triggers a 'close' event, so the above listener will fire
    }
  }, 5*1000).unref();
};

PeerManager.prototype.shutdown = function shutdown() {
  this.state = 'shutdown';
  if (this.server !== false) {
    this.server.close();
  }
  for (var uuid in this.activePeers) {
    if (this.activePeers.hasOwnProperty(uuid) && this.activePeers[uuid] instanceof Peer) {
      this._error('Disconnecting '+uuid, 'notice');
      this._disconnect(this.activePeers[uuid]);
    }
  }
};

// Trigger an error message. Severity is one of 'info', 'notice', 'warning', or 'error' (in increasing severity)
PeerManager.prototype._error = function _error(message, severity) {
  severity = severity || 'warning';
  this.emit('error', {
    severity: severity,
    message: message
  });
};